### DISCUSSION TITLE
gRPC custom client/server certificates & grpc channel options

### DISCUSSION THREAD

**Discussion by Ten0** [Kong/insomnia#3665] [Ideas]
**Is your feature request related to a problem? Please describe.**
Can't use insomnia for gRPC because we authenticate to our gRPC services using a custom PKI, and need to pass particular additional [options](https://grpc.github.io/grpc/node/grpc.Client.html#Client__anchor) to the client builder (`grpc.ssl_target_name_override`, `grpc.keepalive_time_ms`,...).
We are stuck to BloomRPC because of this, and bloom doesn't work super well (lack of ergonomics, environment management, and diverse bugs).

**Describe the solution you'd like**
Being able to configure, ideally by workspace with overrides per environment, the set of CA key, client certificate, client key and [additional options](https://grpc.github.io/grpc/node/grpc.Client.html#Client__anchor).

This is where this should hit in the code:
https://github.com/Kong/insomnia/blob/f27cd5f6306eb50787e31d1f4cfe2606e0c03eb2/packages/insomnia-app/app/network/grpc/index.js#L26-L30

becoming in effect:

```typescript
 const credentials = enableTls ? grpc.credentials.createSsl('ca file content', 'private key file content', 'cert chain file content') : grpc.credentials.createInsecure(); 
  
 console.log(`[gRPC] connecting to url=${url} ${enableTls ? 'with' : 'without'} TLS`); 
 const Client = grpc.makeGenericClientConstructor({}); 
 return new Client(url, credentials, {'grpc.keepalive_time_ms': 12, 'grpc.ssl_target_name_override': 'value' /* and other custom settings*/});
```
(note that it looks like gRPC Core does not receive the options in the same way if they are integers or strings, so this may matter. I'd be very happy with providing these as a JSON, similarly to what's done with the environments.)

**Comment by Ten0** [comment]
For anyone interested I've created some form of dirty patch that provides this functionality.
Feel free to PR a better integration :)

https://github.com/Ten0/insomnia/commits/grpcs_certificates

**Reply by brianjobling** [reply]
Hello @Ten0 this is exactly the feature I was looking for as I want to pass the ssl_target_name_override. I have looked at the code, but don't know where I am meant to put these definitions - can you help?

**Reply by Ten0** [reply]
You want to put `grpc.ssl_target_name_override` in the last options of the client building as hinted [here](https://github.com/Kong/insomnia/discussions/3665#discussion-3439323), so it needs to be put as:
```json
{
    "__grpc_additional_client_options": {
        "grpc__ssl_target_name_override": "value"
    }
}
```
in the insomnia environment. ([1](https://github.com/Ten0/insomnia/commit/8837a8e092cd41cca2d757a053d71a25d711361d#diff-0d876a97684cc6c868835d083785c52894f879ce84216e92bdb0a9876d4abcafR455-R460), [2](https://github.com/Ten0/insomnia/commit/8837a8e092cd41cca2d757a053d71a25d711361d#diff-d3793593bec394b2b1236eb8397a125be94cd7425fa7a1e6e38e82cde1825909R44)).

**Comment by mdeville** [comment]
The SSL certificates support to have been added [here](https://github.com/Kong/insomnia/pull/8235) but not the [custom client gRPC options](https://grpc.github.io/grpc/core/group__grpc__arg__keys.html)
